linesearch.go

```diff
@@ -71,29 +75,30 @@ func (l *Linesearch) Iterate(loc *Location, xNext []float64) (EvaluationType, It
 	}
 	if l.Method.Finished(linesearchLocation) {
 		if l.lastEvalType == FunctionEval {
-			l.finished = true
-			l.finishedF = loc.F
 			// We have the function value at the current location, but we don't
-			// have the gradient, so get it before starting the next major iteration.
+			// have the gradient, so get it before announcing MajorIteration.
+			l.nextIterType = SubIteration
 			copy(xNext, loc.X)
-			return GradientEval, SubIteration, nil
+			return GradientEval, l.nextIterType, nil
 		}
-		return l.initNextLinesearch(loc, xNext)
+		// No gradient evaluation is necessary, announce MajorIteration.
+		l.nextIterType = MajorIteration
+		copy(xNext, loc.X)
+		return NoEvaluation, l.nextIterType, nil
 	}
 
-	// Line search not done, just iterate
+	// Line search not done, just iterate.
 	stepSize, evalType, err := l.Method.Iterate(linesearchLocation)
 	if err != nil {
 		return NoEvaluation, NoIteration, err
 	}
 	floats.AddScaledTo(xNext, l.initLoc, stepSize, l.direction)
 	l.lastEvalType = evalType
-	return evalType, MinorIteration, nil
+	return evalType, l.nextIterType, nil
```

Comment:
I know it's not strictly necessary, but could you have a 
````
l.nextIterType = MinorIteration
````
right after the l.lastEvalType line? It makes it more clear than having to remember it was set in initNextLinesearch and then trying to ensure it didn't get changed somehow.